src/basic/fileio.c

```diff
@@ -1278,3 +1254,97 @@ int fputs_with_space(FILE *f, const char *s, const char *separator, bool *space)
 
         return fputs(s, f);
 }
+
+int open_tmpfile_unlinkable(const char *directory, int flags) {
+        char *p;
+        int fd;
+
+        assert(directory);
+
+        /* Returns an unlinked temporary file that cannot be linked into the file system anymore */
+
+#ifdef O_TMPFILE
+        /* Try O_TMPFILE first, if it is supported */
+        fd = open(directory, flags|O_TMPFILE|O_EXCL, S_IRUSR|S_IWUSR);
+        if (fd >= 0)
+                return fd;
+#endif
+
+        /* Fall back to unguessable name + unlinking */
+        p = strjoina(directory, "/systemd-tmp-XXXXXX");
+
+        fd = mkostemp_safe(p, flags);
+        if (fd < 0)
+                return fd;
+
+        (void) unlink(p);
+
+        return fd;
+}
+
+int open_tmpfile_linkable(const char *target, int flags, char **ret_path) {
+        _cleanup_free_ char *tmp = NULL;
+        int r, fd;
+
+        assert(target);
+        assert(ret_path);
+
+        /* Creates a temporary file, that shall be renamed to "target" later. If possible, this uses O_TMPFILE – in
+         * which case "ret_path" will be returned as NULL. If not possible a the tempoary path name used is returned in
+         * "ret_path". Use link_tmpfile() below to rename the result after writing the file in full. */
+
+#ifdef O_TMPFILE
+        {
+                _cleanup_free_ char *dn = NULL;
+
+                dn = dirname_malloc(target);
+                if (!dn)
+                        return -ENOMEM;
+
+                fd = open(dn, O_TMPFILE|flags, 0640);
```

Comment:
I think we should reject `open_tmpfile_linkable(path, ...|O_EXCL|..., fallback)`. (Or `flags & ~O_EXCL`). This prevents such scenario:
```
fd = open_tmpfile_linkable("./hola", ...|O_EXCL|..., &fallback); (ok)
...
r = link_tmpfile(fd, fallback, "hola"); (hard to debug error on some filesystems)
# linkat(AT_FDCWD, "/proc/self/fd/3", AT_FDCWD, "./hola", AT_SYMLINK_FOLLOW) = -1 ENOENT (No such file or directory)
``` 
> If O_EXCL is not specified, then linkat(2) can be used to link the temporary file into the filesystem, making it permanent,